### PR TITLE
deterministic snapshots

### PR DESCRIPTION
**Pull request by wanderingbort**
Starting this review early. 

This is the plumbing and unit tests for creating:
 - deterministic platform agnostic snapshots of all of the consensus state required to bootstrap a node at a given head block
   - these snapshots are taken of the `HEAD` state of the chain BUT are loaded as if irreversible so, snapshots should only be used if the head block they represent become irreversible until `irreversible` read mode is available.
   - these snapshots can be loaded optionally replaying blocks from a block.log if present and newer than the snapshot
   - the node may continue on perpetually but cannot act as a source of syncing for blocks which are not present in the block.log (it is valid to provide a full block log and a snapshot to just skip the time it takes to build up the state of the chain)
   - the snapshot can be binary or JSON encoded (but its really quite large most of the time)
 - a deterministic platform agnostic integrity hash of the entire chain state (comparable between nodes to check derived state).

Things left to do:
- [x] RPC command to drop the next convenient snapshot
- [x] application level versioning
- [ ] net_plugin review to make sure incomplete block logs don't break anything critical

### configuring a snapshot directory

By default snapshots are written to the `snapshots` directory relative to your `nodeos` data directory.  This can be overridden with a relative or absolute path using the `snapshots_dir` configuration or using the `--snapshots_dir` command line option.  

Snapshots are written out to files named with the pattern `snapshot-<head block id in hex>.bin`

### Triggering the creation of a snapshot

Snapshots can be triggered at runtime using the RPC available through the `producer_api_plugin`.  For example:

```
$ curl http://127.0.0.1:8888/v1/producer/create_snapshot | json_pp
{
   "snapshot_name" : "<data-dir>/snapshots/snapshot-0000002091fd7cb9f1656d25d3a20a93c8aff044c5de19d7cf6f9bab094e3a56.bin",
   "head_block_id" : "0000002091fd7cb9f1656d25d3a20a93c8aff044c5de19d7cf6f9bab094e3a56"
}
``` 

### Instantiating a node from a snapshot

When starting a new instance of `nodeos` the `--snapshot <path to snapshot>` command line option can be used instead of replaying a `blocks.log` or resyncing from a network.  If a `blocks.log` is provided it MUST at least contain blocks up to the snapshotted block and MAY contain additional blocks that will be applied as part of startup.  In addition, any available `reversible/` blocks will be applied.

When instantiating a node from a snapshot, it is illegal to pass in any `--genesis-*` arguments as that information is loaded from a snapshot.  If a `blocks.log` exists, the genesis information it contains will be validated against the genesis data in the snapshot and error if they are not consistent.

***NOTE*** instantiating a node from a snapshot without a blocks.log is valid BUT it will create a partial blocks.log which will affect its ability to service RPC/P2P requests for block data.  

## Instructions for using integrity hashes

Integrity hashes are a way of comparing the contents of the blockchain state database.  They consist of a sha256 hash of a deterministic binary representation of all consensus affecting state.

### calculating an integrity hash

Integrity hashes can be retrieved at runtime using the RPC available through the `producer_api_plugin`.  For example:

```
$ curl http://127.0.0.1:8888/v1/producer/get_integrity_hash | json_pp
{
   "head_block_id" : "00000503289f5f9acd8a8f880b6d91d09817e6806e8f428c6081634a99f3e923",
   "integrity_hash" : "f6c9d3d088e6ceaf87af5e52f8b210a78caf8b831c3b8e890c96c0e6edf734c9"
}
```

### CODE REVIEW

**Comment by wanderingbort** [comment]
@dskvr please see the documentation above and comment on any concerns 😄

File: libraries/chain/block_log.cpp [hunk 1]
```diff
@@ -200,44 +219,50 @@ namespace eosio { namespace chain {
...
+      my->block_stream.write(data.data(), data.size());
       my->genesis_written_to_block_log = true;
 
-      auto ret = append( genesis_block );
+      // append a totem to indicate the division between blocks and header
+      auto totem = npos;
+      my->block_stream.write((char*)&totem, sizeof(totem));
+
+      if (first_block) {
+         auto ret = append(first_block);
```
Comment (heifner) [inline libraries/chain/block_log.cpp]: I would think `ret` would be warning with some compilers.. unused.

**Comment by wanderingbort** [comment]
The original version of the snapshot was not safe WRT foreign keys,  The second version of the snapshot was really slow.  This represents a 3rd attempt at foreign key support that is performant.   Smoke tests on existing chains are coming back positive.

**Comment by wanderingbort** [comment]
depends on EOSIO/fc#29

**Comment by wuyahuang** [comment]
1. Take a snapshot.
2. Instantiating a new instance on previous snapshot.
3. kill the instance and try to restart.
4. "database dirty flag set"

**Comment by XHQ18** [comment]
I get a snapshot in regtest. The file is also generated normally.Then I delete the block.log and other related data files. Start nodeos use --snapshot <path to snapshot> command line option, but failure，why?
$ curl http://127.0.0.1:8888/v1/producer/create_snapshot | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   222  100   222    0     0    520      0 --:--:-- --:--:-- --:--:--   521
{
   "head_block_id" : "0000006eea0a409f6d79139a37fa3f3bc6c21e23b9b4d61a7665b66741daad3e",
   "snapshot_name" : "/Users/xiehanqing/eosbuild2/eos/snapshots/snapshot-0000006eea0a409f6d79139a37fa3f3bc6c21e23b9b4d61a7665b66741daad3e.bin"
}

info  2018-11-02T08:38:31.852 thread-0  chain_plugin.cpp:333          plugin_initialize    ] initializing chain plugin
warn  2018-11-02T08:38:31.852 thread-0  chain_plugin.cpp:684          plugin_initialize    ] 3240000 snapshot_exception: Snapshot exception
Binary snapshot has unexpected magic number!
    {}
    thread-0  snapshot.cpp:212 validate
error 2018-11-02T08:38:31.852 thread-0  main.cpp:130                  main                 ] 3240000 snapshot_exception: Snapshot exception
Binary snapshot has unexpected magic number!
    {}
    thread-0  snapshot.cpp:212 validate
rethrow
    {}
    thread-0  chain_plugin.cpp:684 plugin_initialize

**Comment by XHQ18** [comment]
I get snapshot in own chain.
```
curl http://127.0.0.1:8888/v1/producer/create_snapshot | json_pp
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   220  100   220    0     0    526      0 --:--:-- --:--:-- --:--:--   525
{
   "head_block_id" : "000000e80db33c60505f9a7e012313290a054c38ac376e515ab136d9c858bc6e",
   "snapshot_name" : "/Users/xiehanqing/eoscode/eos/snapshots/snapshot-000000e80db33c60505f9a7e012313290a054c38ac376e515ab136d9c858bc6e.bin"
}
```
Then I delete blocks.log, blocks.index and reversible.
I start nodeos with `--snapshot <path to snapshot>` command line option ,but fail
```
error 2018-11-02T10:42:30.915 thread-0  main.cpp:130                  main                 ] 3240000 snapshot_exception: Snapshot exception Binary snapshot has unexpected magic number!
```
why?

**Comment by XHQ18** [comment]
@wanderingbort Looking forward to your reply

**Comment by ChenJiaoZhu** [comment]
> the snapshot can be binary or JSON encoded
Could you please tell me how to create the snapshot in JSON format? Thanks

**Pull request merged**